### DISCUSSION TITLE
Using MAL to track running percentage

### DISCUSSION THREAD

**Discussion by BFergerson** [apache/skywalking#11062] [Q&A] [answered]
I have a process that creates two meters `junit_test_success` and `junit_test_failed`. I wish to use these two meters to keep a running calculation of a percentage that involves calculating the success rate. That is:
```yaml
exp: "(junit_test_success*100 / (junit_test_success + junit_test_failed)).sum(['service']).service(['service'], Layer.GENERAL)")
```

This works for tracking the success rate for a given day but not for the success rate at a specific time. I would have to manually iterate each record for that day to determine the success rate at that time.

I tried downsampling by sum, latest, and average but none of those functions get me there. `avg` & `sum` have the problem above, and `latest` will either be 100% or 0% (as the meters either send a 1 or 0 to indicate success).

Am I crazy or is the MAL system missing a way to keep track of a running percentage? I know it's possible as it looks like OAL is able to do it. Do I need to introduce a similar function for MAL?

**Comment by wu-sheng** [comment]
I am not following the cause.
Is there you are reportig the metrics not in one bulk?

Could you put the examples of data?

**Reply by BFergerson** [reply]
Here is an example of the data:
![image](https://github.com/apache/skywalking/assets/3278877/45317973-93da-4f7e-a564-44d20d88b3c3)

`junit_test_successful` is a counter that uses the above expression to calculate its value. You can see in the above image that I've updated the counter 4 times with two successes and two failures for a success rate of 50%. That shows up correctly in the hour and day buckets but what I'm looking to do is know what the success rate is per minute. For example, the success rate at `202307100307` should be 33% as at that point in time the test failed twice and succeeded once.

**Reply by wu-sheng** [reply]
I think that is correct from a math perspective. Downsampling is a calculation based on statistics you made. So, it is actually avg of the percent values. 
If you want the final correct result, choose MQE to run calculation in the query stage.

**Reply by BFergerson** [reply]
I'm not sure what you mean by "run calculation in the query stage". Can you explain?

**Reply by wu-sheng** [reply, marked as answer]
https://skywalking.apache.org/docs/main/v9.5.0/en/api/metrics-query-expression/

New query in 9.5.0

**Reply by BFergerson** [reply]
I was unaware of MQE. Thanks, I'll look into this.

**Reply by BFergerson** [reply]
@wu-sheng, MQE allows me to determine the success rate at a given time but how would go about creating a running success rate metric? That is, how would I go about getting the following output:
> ![image](https://github.com/apache/skywalking/assets/3278877/bf54b4fc-836c-4ca2-a071-3cd90c1195a2)
> Two successes followed by two failures

The above shows the current success rate is 50% but at `202307101920` it was at 66.66%.

**Reply by wu-sheng** [reply]
> but how would go about creating a running success rate metric?

Like you had been through, it is nearly impossible unless you use OAL rather than MAL. You never have a chance to get this kind of thing to work.
The only solution is, you use two metrics, one is the success counter, the other is the total counter, and you use `SUM` downsampling. Then get the final result in the query. 
That is the math issue :)

**Comment by wu-sheng** [comment]
And is junit_test_success a counter or gauge?
MAL is statistics oriented, we should make sure it is mathematical correct, before how to write MAL scripts
